src/adc.rs

```diff
@@ -0,0 +1,636 @@
+
+                    // Read 2nd block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw2().clear_bit());
+                    while self.rb.cr.read().lincalrdyw2().bit_is_set() {}
+                    let res_2 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 3rd block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw3().clear_bit());
+                    while self.rb.cr.read().lincalrdyw3().bit_is_set() {}
+                    let res_3 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 4th block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw4().clear_bit());
+                    while self.rb.cr.read().lincalrdyw4().bit_is_set() {}
+                    let res_4 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 5th block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw5().clear_bit());
+                    while self.rb.cr.read().lincalrdyw5().bit_is_set() {}
+                    let res_5 = self.rb.calfact2.read().lincalfact().bits();
+
+                    // Read 6th block of linear correction
+                    self.rb.cr.modify(|_, w| w.lincalrdyw6().clear_bit());
+                    while self.rb.cr.read().lincalrdyw6().bit_is_set() {}
+                    let res_6 = self.rb.calfact2.read().lincalfact().bits();
+
+                    AdcCalLinear([res_1, res_2, res_3, res_4, res_5, res_6])
+                }
+
+                fn check_linear_read_conditions(&self) {
+                    // Ensure the ADC is enabled and is not in deeppowerdown-mode
+                    if self.rb.cr.read().deeppwd().bit_is_set() {
+                        panic!("Cannot read linear calibration value when the ADC is in deeppowerdown-mode");
+                    }
+                    if self.rb.cr.read().advregen().bit_is_clear() {
+                        panic!("Cannot read linear calibration value when the voltage regulator is disabled");
+                    }
+                    if self.rb.cr.read().aden().bit_is_clear() {
+                        panic!("Cannot read linear calibration value when the ADC is disabled");
+                    }
+                }
+
+                fn reset(&mut self, ahb: &mut $AHB) {
+                    ahb.rstr().modify(|_, w| w.$adcxrst().set_bit());
+                    ahb.rstr().modify(|_, w| w.$adcxrst().clear_bit());
+                }
+
+                fn enable_clock(&mut self, ahb: &mut $AHB, d3ccipr: &mut D3CCIPR) {
+                    // Set per_ck as adc clock, TODO: we might want to change this so we can also
+                    // use other clocks as input for this
+                    d3ccipr.kernel_ccip().modify(|_, w| unsafe { w.adcsel().bits(0b10) });
+                    ahb.enr().modify(|_, w| w.$adcxen().set_bit());
```

Comment:
If the adc_ker_ck is out of specification (DS12110 Rev 7 Table 87.), then the ADC will return bad results (with the LSBs always zero, I think you've seen this). Therefore it would be good to add an `assert!` here. For example

```rust
#[cfg(any(
    feature = "stm32h742",
    feature = "stm32h743",
    feature = "stm32h753",
    feature = "stm32h750"
))]
let adc_ker_ck_max = 36_000_000;

assert!(ccdr.clocks.per_ck().unwrap().0 <= adc_ker_ck_max,
        "per_ck is not running or too fast");
```